generic-monitoring/src/main/java/com/powsybl/openrao/monitoring/Monitoring.java

```diff
@@ -0,0 +1,407 @@
+                        applyOptimalRemedialActionsOnContingencyState(state, networkClone, crac, raoResult);
+                        Set<? extends Cnec> currentStateCnecs = physicalParameter.equals(PhysicalParameter.ANGLE) ? crac.getAngleCnecs(state) : crac.getVoltageCnecs(state);
+                        monitoringResult.combine(monitorCnecs(state, currentStateCnecs, networkClone, monitoringInput));
+                        networkPool.releaseUsedNetwork(networkClone);
+                        return null;
+                    })).toList();
+                for (ForkJoinTask<Object> task : tasks) {
+                    try {
+                        task.get();
+                    } catch (ExecutionException e) {
+                        throw new OpenRaoException(e);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            Thread.currentThread().interrupt();
+        }
+
+        BUSINESS_LOGS.info("----- {} monitoring [end]", physicalParameter);
+        monitoringResult.printConstraints().forEach(BUSINESS_LOGS::info);
+        return monitoringResult;
+    }
+
+    private MonitoringResult monitorCnecs(State state, Set<? extends Cnec> cnecs, Network network, MonitoringInput monitoringInput) {
+        BUSINESS_LOGS.info("-- '{}' Monitoring at state '{}' [start]", monitoringInput.getPhysicalParameter(), state);
+        boolean lfSuccess = computeLoadFlow(network);
+        if (!lfSuccess) {
+            return makeResultWhenLoadFlowFails(monitoringInput.getPhysicalParameter(), state, monitoringInput.getCrac());
+        }
+
+        // compute cnec results
+        Set<? extends CnecResult> cnecResults = monitoringInput.getPhysicalParameter().equals(PhysicalParameter.ANGLE) ? computeAngles((Set<AngleCnec>) cnecs, network) : computeVoltages((Set<VoltageCnec>) cnecs, network);
+
+        // Check for threshold overshoot for the voltages/angles of each cnec
+        List<AppliedNetworkActionsResult> appliedNetworkActionsResultList = new ArrayList<>();
+        cnecResults.forEach(cnecResult -> {
+            if (cnecResult.thresholdOvershoot()) {
+                // For Cnecs with overshoot, get associated remedial actions
+                Set<NetworkAction> availableNetworkActions = getNetworkActionsAssociatedToCnec(state, monitoringInput.getCrac(), cnecResult.getCnec(), monitoringInput.getPhysicalParameter());
+                // and apply them
+                AppliedNetworkActionsResult appliedNetworkActionsResult = applyNetworkActions(network, availableNetworkActions, cnecResult.getCnec().getId(), monitoringInput);
+                appliedNetworkActionsResultList.add(appliedNetworkActionsResult);
+            }
+        });
+
+        // if AngleMonitoring : Redispatch to compensate the loss of generation/ load
+        if (monitoringInput.getPhysicalParameter().equals(PhysicalParameter.ANGLE)) {
```

Comment:
not necessary. can be generic if we only check if any of the applied RAs are InjectionSetpoints.
more generally, I think we can make it cleaner / more generic by doing a "balnce" check before & after the monitoring